### ISSUE TITLE
org-paste-subtree inserts empty line above paste

### ISSUE THREAD

**Issue by felixwiemuth** [syl20bnr/spacemacs#16274]
#### Description :octocat:

When pasting a subtree into an empty heading (line starting with some `***`, which can be used to set the level of the pasted tree) with `org-paste-subtree`, an empty line is inserted above the inserted subtree.

I would expect no new line to be inserted, and I think this was the behaviour before I noticed this.

The value of `org-blank-before-new-entry` is `((heading) (plain-list-item))`, which should mean no empty lines. And I think it is this variable which worked before, but stopped working (already in 2020-01, where I noticed the problem).

This does not seem to be a bug in org-mode itself, as it does not happen in plain emacs (tested 2022-09).

In case it is related, and someone takes a look at it, I'd like to mention that it sometimes happens that more then the last cut/yanked subtree is pasted, namely subtree(s) from below the yanked or something that was in the kill ring already (if I recall correctly).

#### Reproduction guide :beetle:
Fresh spacemacs installation with evil mode and minimal packages

Copy (yank) the following
```
* Hello
** World
```
and paste (`org-paste-subtree`) at position "_" here:
```
* One
** Two
** _
```

It does not matter whether in insert mode or navigation mode (using evil).

*Observed behaviour:* :eyes: :broken_heart:
```
* One
** Two
 
** Hello
*** World
```

*Expected behaviour:* :heart: :smile:
```
* One
** Two
** Hello
*** World
```

#### System Info :computer:
- OS: gnu/linux
- Emacs: 29.2
- Spacemacs: 0.999.0
- Spacemacs branch: develop (rev. 2dadee571)
- Graphic display: t
- Running in daemon: nil
- Distribution: spacemacs-base
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(emacs-lisp helm multiple-cursors treemacs)
```

**Comment by smile13241324** [comment]
Hmmm this sounds like a package issue to me when I am honest, can you go ahead and try to reproduce the issue on a vanilla emacs first? I don't think that we have any custom org paste functions implemented.

**Comment by felixwiemuth** [comment]
I had tested it in vanilla emacs (see date above) and it was not reproducible there. But it does happen in a fresh spacemacs installation.

**Comment by fnussbaum** [comment]
`org-paste-subtree` does not insert empty lines, it just skips over any existing lines and pastes the subtree just before the next heading or at the end of the buffer. If `point` is positioned after the asterisks in an otherwise empty heading when invoking it, as in the example above, the asterisks are effectively moved to where the subtree is inserted.
This behaviour was changed in Org 9.3, here is the corresponding changelog entry and part of the corresponding source code:

```org
*** ~org-paste-subtree~ no longer breaks sections

Unless point is at the beginning of a headline, ~org-paste-subtree~
now pastes the tree before the next visible headline.  If you need to
break the section, use ~org-yank~ instead.
```
(from https://github.com/emacs-mirror/emacs/blob/3dfca6f9c7f4da512fff48cf6957c6492e2c0449/etc/ORG-NEWS#L2726)

```elisp
       ;; Remove the forced level indicator.
       (when (and force-level (not level))
         (delete-region (line-beginning-position) (point)))
       ;; Paste before the next visible heading or at end of buffer,
       ;; unless point is at the beginning of a headline.
       (unless (and (bolp) (org-at-heading-p))
         (org-next-visible-heading 1)
         (unless (bolp) (insert "\n")))
```
So the behaviour observed above could be due either to a mismatch in the Org versions, or to the presence or absence of newline characters at the end.
In my tests, the behaviour in Spacemacs seemed to be consistent with `emacs -Q`.

**Comment by felixwiemuth** [comment]
Thanks for sharing this observation and testing again with plain emacs.
Regarding that 9.3 was release in 12/2019, there is a time-wise relation as I discovered the issue in 01/2020.

However, regarding that I tested with plain emacs 2022, I don't think this was due to an old version.
`emacs -Q` is somehow not reliably having the `org-paste-subtree` command (it suddenly was available after some time), and copy or yank and then `org-paste-subtree` fails ("is not a tree").

Regarding a newline character after `point` where pasting, this indeed makes a difference (tested in Spacemacs, orgmode 9.6.15):
- if there is one, it ends up with *two* newlines instead of one as in my original example
- without one, it is like in the original example.

It is enough that `point` is on a single space after the `**` for the newline to occur after pasting, whether that space ends the line or more spaces are following. Basically it seems that all whitespace after asterisk (whether point is on asterisk or whitespace) is jumped over and put on a line between `** Two` and the pasted subtree.
- If `point` is on a whitespace, only the whitespace from there to the end of the line is retained
- If `point` is on the asterisk, all whitespace after the asterisk is retained but becomes a same-level headline (same if there is non-whitespace following the whitespace)

**Comment by fnussbaum** [comment]
Thanks, I can reproduce your observations, they helped me realise what the problem might be. Although the behaviour is consistent with the docstring of `org-paste-subtree`, it might be a little bit inconvenient when using evil-mode. For me, the behaviour in `evil-insert-state` is exactly the same as with `emacs -Q`.
The behaviour in `evil-normal-state` is also the same, but here we have the problem that the cursor moves back one character by default when leaving normal mode. If there is a line `**_` at the end of the buffer (in particular, no final newline, which is often automatically inserted) where the underscore marks a space, and the cursor is on the space in normal mode, then `org-paste-subtree` behaves exactly like it does in insert mode (at the same position, i.e. `**|_`). What happens is this: The asterisks are deleted, then `(org-next-visible-heading 1)` is executed, which in this case moves point to the end of the buffer, which is just after the space. Then `(bolp)` returns `nil` and a new line character is inserted and finally the subtree is pasted at the appropriate level (compare the code snippet I posted above). Thus the space remains lonely on the original line.

A way to deal with this could be either calling `org-paste-subtree` from insert mode via `M-m m s p` (where the cursor may be at the end of the line after the asterisks, or after any spaces), or more drastically: Set `evil-move-cursor-back` to `nil` and `evil-move-beyond-eol` to `t`.
Maybe the most convenient way would be to tweak the behaviour of `org-paste-subtree` as desired by using a custom wrapping function (or advice).

>     * If `point` is on the asterisk, all whitespace after the asterisk is retained but becomes a same-level headline (same if there is non-whitespace following the whitespace)

This also seems consistent with the docstring and the behaviour in insert mode or `emacs -Q`: It behaves as if, in `emacs -Q`, one inserted some asterisks and then moved the cursor back one character.

**Comment by fnussbaum** [comment]
Thanks, I can reproduce your observations. For me, the behaviour in Spacemacs and `emacs -Q` seems consistent, there is only the following difference due to evil-mode: When leaving `evil-insert-state`, the cursor moves back one character by default. See the variables `evil-move-cursor-back` and `evil-move-beyond-eol`. I included this observation previously in another comment, but deleted it now as it did not address the crux of the matter.

I found a [thread](https://list.orgmode.org/orgmode/a2cc49d3-7f65-5f6a-8f05-21e5056335b3@gmail.com/) corresponding to your issue on the Org mailing list, and it has just been fixed. In my tests it also works well with the default behaviours of the evil cursor. I would expect the fix to be included in the next Org release, presumably 9.6.28.
